### DISCUSSION TITLE
The Path for Testing Path Traversal Vulnerabilities with Httpx

### DISCUSSION THREAD

**Discussion by C1ph3rX13** [encode/httpx#2897] [Q&A]
I have noticed an odd behavior in the httpx module. 

import requests
httpx.get("http://127.0.0.1/../../../../replay")

Which should typically send the request to the following path: /../../../../replay. Instead, it requests "http://127.0.0.1/replay" and modifies the specified input.

But I want to use the httpx to send HTTP requests but do not want the URL to be modified.

**Comment by lovelydinosaur** [comment]
Yo. 🤘

Sure, `httpx` handles path traversal which is what we'd typically want.
Same as browser behaviour, and also `requests`.

If you want to break past that constraint a good place to start would be [working with `httpcore` directly](https://www.encode.io/httpcore/).

Taking a look at the `HTTPTransport` class is also instructive if you want to dig into the design here... https://github.com/encode/httpx/blob/e63b6594f2863b7c8274eb0991ebc6cad63661f7/httpx/_transports/default.py#L215-L222

...if we wanted to also expose similar functionality in `httpx` we'd need a way of allowing URLs to be constructed with an explicitly unescaped `target` specified.

**Reply by C1ph3rX13** [reply]
with requests.Session() as s:
            prep_resp = requests.Request(method='GET', url=replay_url, headers=headers)
            prep = prep_resp.prepare()
            prep.url = replay_url
            download_resp = s.send(prep, verify=False)

A way of allowing URLs to be constructed with an explicitly unescaped target specified by using `request`

It would be great if httpx could implement this function

**Comment by berezhinskiy** [comment]
@C1ph3rX13 Have you found any technique to preserve the exact request-target with ../ using httpx without writing a custom transport?  Any pointers or examples would be hugely appreciated.

**Reply by zrquan** [reply]
Try [target extension](https://www.python-httpx.org/advanced/extensions/#target).

```python
httpx.Client().get("http://localhost:8000", extensions={"target": "/../../../api"})
```

**Reply by berezhinskiy** [reply]
Thank you so much. This works perfectly!
